File: 2192. All Ancestors of a Node in a Directed Acyclic Graph.java
class Solution {
    public List<List<Integer>> getAncestors(int n, int[][] edges) {
        List<Integer>[] graph = new ArrayList[n];
        List<List<Integer>> result = new ArrayList<>();
        for (int i = 0; i < graph.length; i++) {
            graph[i] = new ArrayList<>();
            result.add(new ArrayList<>());
        }
        for (int[] edge : edges) {
            graph[edge[0]].add(edge[1]);
        }
        for (int i = 0; i < n; i++) {
            dfs(i, i, graph, result);
        }
        return result;
    }
    private void dfs(int current, int ancestor, List<Integer>[] graph, List<List<Integer>> result) {
        for (int child : graph[current]) {
            if (result.get(child).size() == 0 || result.get(child).get(result.get(child).size()-1) != ancestor) {
                result.get(child).add(ancestor);
                dfs(child, ancestor, graph, result);
            }
        }
    }
}

File: 2192. All Ancestors of a Node in a Directed Acyclic Graph.go
func getAncestors(n int, edges [][]int) [][]int {
    result := make([][]int, n)
    adj := make([][]int, n)
    for i := range adj {
        adj[i] = []int{}
        result[i] = []int{}
    }
    for _, edge := range edges {
        from := edge[0]
        to := edge[1]
        adj[from] = append(adj[from], to)
    }
    for i := 0; i < n; i++ {
        dfs(i, i, adj, result)
    }
    return result
}

func dfs(current, ancestor int, graph, result [][]int) {
    for _, child := range graph[current] {
        if len(result[child]) == 0 || result[child][len(result[child])-1] != ancestor {
            result[child] = append(result[child], ancestor)
            dfs(child, ancestor, graph, result)
        }
    }
}

